Make it possible to run nfd as a DaemonSet

marquiz opened the pull request:
The option to run NFD as a one-shot job is still available

balajismaniam commented on main.go:
```diff
@@ -70,39 +72,63 @@ type APIHelpers interface {
+			break
+		}
+
+		if args.sleepInterval > 0 {
```
I don't we should allow too low values for sleep interval. If the value is too low, the DaemonSet pods might go into a crash loop back off state. WDYT?

marquiz replied:
As I commented earlier, it makes sense to restrict the sleep interval to some sane value. However, I would change it so that too small but non-zero values are forced to the minimum and a warning is printed. Zero value is used to specify an infinite sleep (i.e. no re-labeling).

balajismaniam commented on main.go:
```diff
@@ -70,39 +72,63 @@ type APIHelpers interface {
+			stderrLogger.Fatalf("error occurred while updating node with feature labels: %s", err.Error())
+		}
+
+		if args.oneshot == true {
```
s/args.oneshot == true/args.oneshot

marquiz replied:
ok

balajismaniam commented on main.go:
```diff
@@ -152,21 +185,21 @@ func configureParameters(sourcesArg []string, whiteListArg string) (sources []so
+	labelWhiteList, err = regexp.Compile(labelWhiteListStr)
 	if err != nil {
-		stderrLogger.Printf("error parsing whitelist regex (%s): %s", whiteListArg, err)
+		stderrLogger.Printf("error parsing whitelist regex (%s): %s", labelWhiteList, err)
```
s/labelWhiteList/labelWhiteListStr

marquiz replied:
Uh, thanks for spotting this

balajismaniam commented on main.go:
```diff
@@ -125,19 +155,22 @@ func argsParse(argv []string) (noPublish bool, sourcesArg []string, whiteListArg
-	enabledSources := map[string]struct{}{}
-	for _, s := range sourcesArg {
-		enabledSources[strings.TrimSpace(s)] = struct{}{}
+func configureParameters(sourcesWhiteList []string, labelWhiteListStr string) (enabledSources []source.FeatureSource, labelWhiteList *regexp.Regexp, err error) {
```
Are these changes to `configureParameters()` required?

marquiz replied:
I tried to make the variable naming a bit more comprehensible/self-explanatory. The current naming confused me when reading the code for the first time.

marquiz commented on README.md:
```diff
@@ -144,8 +144,16 @@ such as restricting discovered features with the --label-whitelist option._
-Feature discovery is done as a one-shot job. There is an example script in this
-repo that demonstrates how to deploy the job to unlabeled nodes.
+Feature discovery is preferably run as a Kubernetes DaemonSet. There is an
+example spec that you can use as a template, or, as is for trying out the
```
ok

marquiz commented on main.go:
```diff
@@ -70,39 +72,63 @@ type APIHelpers interface {
+	labelWhiteList string
+	noPublish      bool
+	oneshot        bool
+	sleepInterval  int
```
Makes sense, I wasn't aware of ParseDuration

marquiz commented on main.go:
```diff
@@ -70,39 +72,63 @@ type APIHelpers interface {
   Usage:
-  %s [--no-publish --sources=<sources> --label-whitelist=<pattern>]
+  %s [--no-publish] [--sources=<sources>] [--label-whitelist=<pattern>]
+        [--oneshot | --sleep-interval=<seconds>]
```
Hmm, thanks for spotting this. I think this was just a "typo" from me

marquiz commented on main.go:
```diff
@@ -125,19 +155,22 @@ func argsParse(argv []string) (noPublish bool, sourcesArg []string, whiteListArg
+	args.sources = strings.Split(arguments["--sources"].(string), ",")
+	args.labelWhiteList = arguments["--label-whitelist"].(string)
+	args.oneshot = arguments["--oneshot"].(bool)
+	args.sleepInterval, _ = strconv.Atoi(arguments["--sleep-interval"].(string))
```
It shouldn't end up in crashloopbackoff because it's a daemon, i.e. it doesn't exit, right(?) When using parseduration it probably makes sense to restrict it to some sane value, e.g. 1 second

balajismaniam replied:
I misread the code last time. It should not crashloopbackoff here. Thanks.

marquiz commented on main.go:
```diff
@@ -114,7 +140,11 @@ func argsParse(argv []string) (noPublish bool, sourcesArg []string, whiteListArg
+  --oneshot                   Label once and exit.
+  --sleep-interval=<seconds>  Number of seconds to sleep between re-labeling.
+                              Non-positive value implies no re-labeling (i.e.
+                              infinite sleep). [Default: 60]`,
```
Yup

marquiz commented:
I pushed a new version, addressing the comments from @balajismaniam. I didn't change the variable renaming in configureParameters(), i.e. it is as it was in the first version of the patchset.

marquiz commented:
> Thanks for working on this @marquiz. Can you add a description of how you tested this PR end-to-end for posterity?

1. Set-up a cluster with multiple nodes
2. Clone nfd and build the docker image
```
$ git clone https://github.com/kubernetes-incubator/node-feature-discovery && cd node-feature-discovery
$ make docker
```
3. Tag and push the docker image to a registry available for all nodes
```
$ docker tag <id> <registry>:<tag>
$ docker push <registry>:<tag>
```
4. Modify the provided template to use the just-build docker image
```
$ sed s'/"image".*/"image": "<registry>:<tag>",/' node-feature-discovery-daemonset.json.template > node-feature-discovery-daemonset.json
$ kubectl apply -f node-feature-discovery-daemonset.json
```
5, After a while, check that the daemonset is healthy and that labels have been set correctly
```
$ kubectl get ds node-feature-discovery
NAME                     DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
node-feature-discovery   3         3         3         3            3           <none>          1m
$ kubectl get nodes  -o yaml | grep 'node.alpha.kubernetes-incubator.io/'
      node.alpha.kubernetes-incubator.io/nfd-cpuid-AESNI: "true"
      node.alpha.kubernetes-incubator.io/nfd-cpuid-AVX: "true"
...
```
6. Add a new node to the cluster
```
$ ssh <new-node>
$ sudo kubeadm join...
$ exit
```
7. Check that the new node gets labeled, too
```
$ kubectl get node <new-node> -o yaml | grep 'node.alpha.kubernetes-incubator.io/'
      node.alpha.kubernetes-incubator.io/nfd-cpuid-AESNI: "true"
      node.alpha.kubernetes-incubator.io/nfd-cpuid-AVX: "true"
...
```

balajismaniam commented on README.md:
````diff
@@ -144,8 +144,16 @@ such as restricting discovered features with the --label-whitelist option._
+service:
+```
+kubectl create -f node-feature-discovery-daemonset.json.template
+```
````
Can you add a blurb on how to change the sleep interval? Something along the lines of:
`When run as a daemonset, the node is relabelled at an interval specified using the --sleep-interval flag. The default interval is set to 60s.`

May be we can add this arg in the daemonset template and add a link to the line in the above documentation as well?

marquiz replied:
> add a link to the line in the above documentation as well?

JSON does not allow comments, afaik

balajismaniam commented on node-feature-discovery-daemonset.json.template:
```diff
@@ -0,0 +1,43 @@
+              }
+            ],
+            "image": "quay.io/kubernetes_incubator/node-feature-discovery:v0.1.0",
+            "name": "node-feature-discovery"
```
May be we should add the `--sleep-interval` arg in this template to show it as an example? This can also be linked in the doc?

balajismaniam commented:
@marquiz Thanks for adding the details on how you tested it. I've left some minor comments on adding docs on `--sleep-interval`. LGTM after that.

marquiz commented:
I changed the documentation as you suggested. However, linking from the template to README is not possible when the template format is JSON. Please take a look at #108 which would ﬁx this problem.

Another change is that I also updated the "Command line interface" section of README to include the new command line options.

balajismaniam commented:
Thanks for making the changes @marquiz. May be the comment was confusing but I meant a link in the readme to the line in the template similar to https://github.com/kubernetes-incubator/node-feature-discovery/blob/master/node-feature-discovery-job.json.template#L34. We don't need #108 for this.

balajismaniam commented on README.md:
````diff
@@ -144,8 +149,20 @@ such as restricting discovered features with the --label-whitelist option._
+```
+
+When run as a daemonset, nodes are re-labeled at an interval specified using
+the --sleep-interval option. In the template the default interval is set to 60s
````
s/--sleep-interval/`--sleep-interval`. Use backtick.

balajismaniam commented on README.md:
```diff
@@ -144,8 +149,20 @@ such as restricting discovered features with the --label-whitelist option._
+
+When run as a daemonset, nodes are re-labeled at an interval specified using
+the --sleep-interval option. In the template the default interval is set to 60s
+which is also the default when no --sleep-interval is specified.
```
s/--sleep-interval/`--sleep-interval`. Use backtick.

balajismaniam replied:
Please link https://github.com/kubernetes-incubator/node-feature-discovery/blob/master/node-feature-discovery-job.json.template#L38 for `--sleep-interval`.

marquiz commented:
I pushed a new version addressing your comments.

> We don't need #108 for this.

OK, but I'm waiting for comments on that, anyway :)
